### DISCUSSION TITLE
(Undo / History) Undo, Spacebar, Lasso/Selection tool, shortcut problems

### DISCUSSION THREAD

**Discussion by GFred6576** [opentoonz/opentoonz#6393] [General]
### Is your feature request related to a problem? Please describe

#1-1 the problems encountered in Otz：when ctrl z ，the range is disappeared
![image](https://github.com/opentoonz/opentoonz/assets/104918260/0bc6f12a-34e3-43f6-b7cc-8f148b782f93)
![image](https://github.com/opentoonz/opentoonz/assets/104918260/7dd986f2-6918-4a22-bd98-2a31982400b3)
![image](https://github.com/opentoonz/opentoonz/assets/104918260/cf23799e-2d8e-483e-8cce-2192635660e1)
![image](https://github.com/opentoonz/opentoonz/assets/104918260/3600727d-2377-4a8c-bb6a-484149a45d7c)
*When I press ctrl z , the selection is gone.
#1-2 the problems encountered in Otz：
![image](https://github.com/opentoonz/opentoonz/assets/104918260/53e40c97-0cbc-4185-961e-939699d35c3d)
![image](https://github.com/opentoonz/opentoonz/assets/104918260/34bbadfa-921d-483f-ab07-ca7746637311)
![image](https://github.com/opentoonz/opentoonz/assets/104918260/f3c929ee-216b-406f-9692-6fa039a2d311)
*When I press space, the range disappears

#2-2 of the problems encountered in Otz：
![image](https://github.com/opentoonz/opentoonz/assets/104918260/92ab3df5-4b88-4ba8-9c67-5b7d4ab79ecc)
![image](https://github.com/opentoonz/opentoonz/assets/104918260/d2d42a69-5692-44d0-8293-c3f324ba056a)
*The problem is, all the layers don't show up,
I thought maybe we could add the ability to toggle Show/Hide the current layer?

![image](https://github.com/opentoonz/opentoonz/assets/104918260/a33ecc31-2cf2-4089-9c3d-47da6ecdc2dd)
*If possible, I would be able to set the shortcut keys in the same way as this function, which I think would be helpful for checking if the wrong layers are drawn

#3
![image](https://github.com/opentoonz/opentoonz/assets/104918260/c978d1b3-8041-4646-847b-8641205c5d6b)

When I press ctrl z or ctrl y all the time, it will pop up. Can I turn it off? Other animation software does not show this, this window feels less needed, and sometimes can be a little annoying

### Describe the solution you'd like

#1 
  I wish I could improve this, for example: currently when I press "space", it turns into the space tool, which moves the canvas around, but it's more like a tool, which I'm not used to. 
  Other animation programs, such as csp and krita, that are currently using the Brush Tool will not switch to the Space tool when the space bar is pressed, but will not affect the currently selected tool. 
  Considering the way the current version is going to work for someone who might like it, maybe add a feature to the preferences that toggles the way the "space" bar is selected?

#2
I thought maybe we could add the ability to toggle Show/Hide the current layer?

#3
This window is "Critical Warning: No more operations available"  Maybe it can be turn off?

**Comment by RodneyBaker** [comment]
Regarding No. 2
We can show/hide the current level via the column icon (columns... rows in the timeline... are basically layers)
There is a "Toggle Show/Hide" that can be given a shortcut key that works well for whatever is selected at the time.

Regarding No. 3
If possible I think it would be more ideal for that Undo process to cycle back to the top again although I can see where users might not like that.   But if cyclic all we'd need to do to move back to that prior location on the Undo stack would be to involke a new Redo. 

Aside:  We can create a history scroller  (undo/redo slider) using Custom Panels and that's where the cycling of history would be particularly useful.

Regarding No. 1   I think I understand the problem but would need to look at that more to be able to comment.
I'm sure there are a number of things that can improve the Move Tool.

**Comment by flurick** [comment]
1.1 Interesting, since selections are not actions that can be undone it's instead just removed with the first transformation. If we remove that one line, the selection would stay until a new selection was made. https://github.com/opentoonz/opentoonz/blob/0c59f70c2d5b05e7f8b668ea7963dc49764ae274/toonz/sources/tnztools/rasterselection.cpp#L1062

1.2 (https://github.com/opentoonz/opentoonz/issues/4900) I have a WIP branch for this but it feels like maybe a user would want to bind the space key to something else, which would not be possible if we lock it to the same function as "mouse wheel panning". Or maybe people just don't remap their space key anyway?
https://github.com/opentoonz/opentoonz/compare/master...flurick:opentoonz:fix-spacebar-tmp-tool-reset

2 Apparently the columns "header" cell is deselected when changing tools in the canvas area, so I only got half way on my first attempt to implement a new "Toggle selected column" command. More code needed to use the active cell if no column "header" is selected.
https://github.com/opentoonz/opentoonz/blob/0c59f70c2d5b05e7f8b668ea7963dc49764ae274/toonz/sources/toonz/columncommand.cpp#L1533

3 I think I agree, they just are one line each that could be removed - or maybe a new kind of warning popup could be used that doesn't need to be closed to continue work? 
https://github.com/opentoonz/opentoonz/blob/0c59f70c2d5b05e7f8b668ea7963dc49764ae274/toonz/sources/toonz/mainwindow.cpp#L936


Some related Tahoma patches:
https://github.com/tahoma2d/tahoma2d/pull/37
https://github.com/tahoma2d/tahoma2d/pull/588
https://github.com/tahoma2d/tahoma2d/pull/583

**Comment by flurick** [comment]
> I think this is another instance of Opentoonz expecting a different workflow. A workflow where traditional drawings are expected to either come in fully planned and ready to be digitally inked and painted via scanning.
> 
> Or a workflow that is paperless, where one does the roughs in one column, and then clean them up in another column.
> 
> Where selecting and moving parts is not expected to happen often, and if it happens, then its expected to be minimal, and not as an integral part of the digital drafting process.
> 
> Otherwise I just cannot add the longevity of toonz together with how utterly basic the drawing tools are. It's not like toonz is a young product that was just invented a short time ago...it's been around for some time. Yet it never needed to evolve better drawing tools.
> 
> So, in my opinion, toonz, and OT just expect a certain way of working.
> 
> Until this is addressed (although I am kind of against it because I like specialist software, especially now that there are many exchange formats to make communication between differenet programs easier) perhaps draw your roughs in a program that has better, more comfortable painting, and then clean them up in OT when the drawings are at a stage that doesn't need this kind of operation anymore.
> 
> Edit: Of course I agree that the selection should not disappear by panning the view with the spacebar. I'm not against innovation, but there is a point where I think: It's easier to just use a more specialized program to do this, and then use OT for timing, compositing, and other animation tasks.
> 
> With no need to recreate Krita in it. And davinci resolve. And blender. And audacity. Etc. These things all take development power and need to be implemented well, and bug free. A massive effort just to avoid a shared folder between Krita and OT and managing some files between them... But I think I'm a minority on that thinking so ...yeah, lol. Again, I agree...but only to an extent.

It's a good point, how much do we actually want in OT and what should be left to other programs. The stuff that _could_ be added for cutout and OpenCV is endless

**Comment by RodneyBaker** [comment]
While there is a lot more going on in this report than just space key related issues...

Recommend reviewing this report in light of #6199 being merged (soon).

**Comment by RodneyBaker** [comment]
Converting this to Discussion so others can explore information and consider the best way forward.
